src/main/kotlin/io/emeraldpay/dshackle/upstream/Selector.kt

```
@@ -693,4 +734,110 @@ class Selector {
 
         override fun toString(): String = "Matcher: ${describeInternal()}"
     }
+
+    class ExactVersionMatcher(
+        private val expectedRawVersion: String,
+    ) : LabelSelectorMatcher() {
+
+        override fun matchesWithCause(labels: UpstreamsConfig.Labels): MatchesResponse {
+            val actualRawVersion = labels["client_version"] ?: return ExactVersionResponse(expectedRawVersion)
+
+            return try {
+                val actual = Semver(actualRawVersion.removePrefix("v"), Semver.SemverType.STRICT)
+                val expected = Semver(expectedRawVersion.removePrefix("v"), Semver.SemverType.STRICT)
+                if (actual.isEquivalentTo(expected)) Success else ExactVersionResponse(expectedRawVersion)
+            } catch (_: SemverException) {
+                if (actualRawVersion == expectedRawVersion) Success else ExactVersionResponse(expectedRawVersion)
+            }
+        }
+
+        override fun asProto(): BlockchainOuterClass.Selector =
+            BlockchainOuterClass.Selector.newBuilder().setExactVersionSelector(
+                BlockchainOuterClass.ExactVersionSelector.newBuilder()
+                    .setVersion(expectedRawVersion)
+                    .build(),
+            ).build()
+
+        override fun describeInternal(): String =
+            "exact version '$expectedRawVersion'"
+
+        override fun toString(): String =
+            "Matcher: ${describeInternal()}"
+    }
+
+    class RangeVersionMatcher(
+        private val minRawVersion: String,
+        private val maxRawVersion: String,
+    ) : LabelSelectorMatcher() {
+
+        override fun matchesWithCause(labels: UpstreamsConfig.Labels): MatchesResponse {
+            val actualRawVersion = labels["client_version"]
+                ?: return RangeVersionResponse(minRawVersion, maxRawVersion)
+
+            val actualSemver = try {
+                Semver(actualRawVersion.removePrefix("v"), Semver.SemverType.STRICT)
```

We can use `runCatching` in all places since we ignore an exception